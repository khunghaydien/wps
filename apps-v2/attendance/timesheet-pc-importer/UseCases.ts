import UseCaseCollection from '@attendance/application/UseCaseCollection';
import { IUseCase as IImporterTimesheetSaveUseCase } from '@attendance/domain/combinedUseCases/importer/ICheckAndSaveTimesheetUseCase';
import { IUseCase as IDailyRecordFetchRestTimeReasonsUseCase } from '@attendance/domain/useCases/dailyRecord/IFetchRestTimeReasonsUseCase';
import { IUseCase as IDailyRequestEarlyLeaveRequestFetchReasonsUseCase } from '@attendance/domain/useCases/dailyRequest/earlyLeaveRequest/IFetchReasonsUseCase';
import { IUseCase as IDailyRequestLateArrivalRequestFetchReasonsUseCase } from '@attendance/domain/useCases/dailyRequest/lateArrivalRequest/IFetchReasonsUseCase';
import { IUseCase as IGeneralFetchUserSettingUseCase } from '@attendance/domain/useCases/general/IFetchUserSettingUseCase';
import { IUseCase as IImporterTimesheetCheckUseCase } from '@attendance/domain/useCases/importer/timesheet/ICheckUseCase';
import { IUseCase as IFetchContractedWorkTimesUseCase } from '@attendance/domain/useCases/importer/timesheet/IFetchContractedWorkTimesUseCase';
import { IUseCase as ILeaveFetchListUseCase } from '@attendance/domain/useCases/leave/IFetchListUseCase';

export type UseCases = {
  fetchUserSetting: IGeneralFetchUserSettingUseCase;
  fetchRestTimeReasons: IDailyRecordFetchRestTimeReasonsUseCase;
  fetchRestTimeReasonsForBulk: IDailyRecordFetchRestTimeReasonsUseCase;
  fetchLeaves: ILeaveFetchListUseCase;
  fetchEarlyLeaveReasons: IDailyRequestEarlyLeaveRequestFetchReasonsUseCase;
  fetchLateArrivalReasons: IDailyRequestLateArrivalRequestFetchReasonsUseCase;
  fetchContractedWorkTimes: IFetchContractedWorkTimesUseCase;
  saveTimesheet: IImporterTimesheetSaveUseCase;
  checkTimesheet: IImporterTimesheetCheckUseCase;
};

const service = UseCaseCollection<UseCases>(
  "Timesheet-PC-importer's UseCase Service"
);

export default service;
