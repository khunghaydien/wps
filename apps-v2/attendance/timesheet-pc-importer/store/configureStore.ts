/* eslint-disable global-require, import/newline-after-import */
import { applyMiddleware, compose, createStore, Middleware } from 'redux';
import logger from 'redux-logger';
// 本来は使いたくないが
// Global Container の社員切替が使用しているので
// しかたなく導入している
import thunkMiddleware from 'redux-thunk';

import rootReducer, { State } from '../modules';

const PRODUCTION = process.env.NODE_ENV === 'production';
const TEST = process.env.TEST === 'test';

/* eslint-disable no-underscore-dangle */
const composeEnhancers =
  !PRODUCTION && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    : compose;

export default function configureStore(initialState?: State) {
  const middleWares: Middleware[] = [
    // 本来は使いたくないが
    // Global Container の社員切替が使用しているので
    // しかたなく導入している
    thunkMiddleware,
  ];
  if (!PRODUCTION && !TEST) {
    middleWares.push(logger);
  }
  const store = createStore(
    rootReducer,
    initialState, // Redux DevTools Extension
    composeEnhancers(applyMiddleware(...middleWares))
  );

  if (!PRODUCTION && module.hot) {
    module.hot.accept('../modules', () => {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const newRootReducer = require('../modules').default;
      store.replaceReducer(newRootReducer);
    });
  }

  return store;
}
