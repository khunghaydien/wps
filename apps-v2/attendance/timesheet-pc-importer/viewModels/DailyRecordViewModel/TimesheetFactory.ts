import { CODE } from '@attendance/domain/models/AttDailyRequestType';
import * as DailyRecord from '@attendance/domain/models/importer/DailyRecord';
import * as Timesheet from '@attendance/domain/models/importer/Timesheet';

import * as DailyRecordViewModel from '.';

const createRestTime = (
  record: DailyRecordViewModel.DailyRecordViewModel
): DailyRecord.DailyRecord['restTimes'] => {
  const restTime: DailyRecord.DailyRecord['restTimes'] = [];
  if (
    (record.rest1StartTime ?? null) !== null &&
    (record.rest1EndTime ?? null) !== null
  ) {
    restTime.push({
      startTime: record.rest1StartTime,
      endTime: record.rest1EndTime,
      restReason: {
        id: '',
        code: record.rest1ReasonCode,
        name: '',
      },
    });
  }
  if (
    (record.rest2StartTime ?? null) !== null &&
    (record.rest2EndTime ?? null) !== null
  ) {
    restTime.push({
      startTime: record.rest2StartTime,
      endTime: record.rest2EndTime,
      restReason: {
        id: '',
        code: record.rest2ReasonCode,
        name: '',
      },
    });
  }
  return restTime;
};

const creteRequests = (
  record: DailyRecordViewModel.DailyRecordViewModel
): DailyRecord.DailyRecord['requests'] => {
  const requests: DailyRecord.DailyRecord['requests'] = [];
  const base = {
    id: null,
    requestTypeName: null,
    status: null,
  };
  if (record.appliedLeaveRequest1) {
    requests.push({
      ...base,
      type: CODE.Leave,
      startDate: record.recordDate,
      endDate: record.recordDate,
      startTime: record.leaveRequest1StartTime,
      endTime: record.leaveRequest1EndTime,
      leaveCode: record.leaveRequest1Code,
      leaveName: '',
      leaveType: null,
      leaveRange: record.leaveRequest1Range,
      remarks: record.leaveRequest1Remark,
      reason: record.leaveRequest1Reason,
      requireReason: false,
    });
  }
  if (record.appliedOvertimeWorkRequest) {
    requests.push({
      ...base,
      type: CODE.OvertimeWork,
      targetDate: record.recordDate,
      startTime: record.overtimeWorkRequestStartTime,
      endTime: record.overtimeWorkRequestEndTime,
      remarks: record.overtimeWorkRequestRemark,
    });
  }
  if (record.appliedEarlyStartWorkRequest) {
    requests.push({
      ...base,
      type: CODE.EarlyStartWork,
      targetDate: record.recordDate,
      startTime: record.earlyStartWorkRequestStartTime,
      endTime: record.earlyStartWorkRequestEndTime,
      remarks: record.earlyStartWorkRequestRemark,
    });
  }
  if (record.appliedEarlyLeaveRequest) {
    requests.push({
      ...base,
      type: CODE.EarlyLeave,
      targetDate: record.recordDate,
      startTime: null,
      endTime: null,
      reasonText: record.earlyLeaveRequestReasonText,
      reasonCode: record.earlyLeaveRequestReasonCode,
    });
  }
  if (record.appliedLateArrivalRequest) {
    requests.push({
      ...base,
      type: CODE.LateArrival,
      targetDate: record.recordDate,
      startTime: null,
      endTime: null,
      reasonText: record.lateArrivalRequestReasonText,
      reasonCode: record.lateArrivalRequestReasonCode,
    });
  }
  if (record.appliedHolidayWorkRequest) {
    requests.push({
      ...base,
      type: CODE.HolidayWork,
      targetDate: record.recordDate,
      startTime: record.holidayWorkRequestStartTime,
      endTime: record.holidayWorkRequestEndTime,
      substituteLeaveType: record.holidayWorkRequestSubstituteLeaveType,
      substituteDate: null,
      reason: null,
      remark: record.holidayWorkRequestRemark,
    });
  }
  return requests;
};

export default {
  create: ({ employeeId, records: $records }) => {
    const records = DailyRecordViewModel.filterForSubmitting($records).map(
      (record): DailyRecord.DailyRecord => ({
        id: '',
        recordDate: record.recordDate,
        startTime: record.startTime,
        endTime: record.endTime,
        restTimes: createRestTime(record),
        requests: creteRequests(record),
        comment: record.comment,
      })
    );
    return {
      employeeId,
      startDate: records.at(0)?.recordDate || null,
      endDate: records.at(-1)?.recordDate || null,
      records,
    };
  },
} as Timesheet.ITimesheetFactory<{
  employeeId: string;
  records: DailyRecordViewModel.DailyRecordViewModel[];
}>;
