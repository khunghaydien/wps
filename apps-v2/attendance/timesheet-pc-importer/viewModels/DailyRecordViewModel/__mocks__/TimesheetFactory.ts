import * as Timesheet from '@attendance/domain/models/importer/Timesheet';

import * as DailyRecordViewModel from '..';

export default {
  create: jest.fn(),
} as Timesheet.ITimesheetFactory<DailyRecordViewModel.DailyRecordViewModel[]>;
