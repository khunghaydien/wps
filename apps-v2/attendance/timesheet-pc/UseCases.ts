import UseCaseCollection from '@attendance/application/UseCaseCollection';
import { IUseCase as ISubmitFixDailyRequestAndSaveDailyRecordUseCase } from '@attendance/domain/combinedUseCases/ISaveDailyRecordAndSubmitFixDailyRequestUseCase';
import { IUseCase as ISubmitFixDailyRequestWithClockOutUseCase } from '@attendance/domain/combinedUseCases/ISubmitFixDailyRequestWithClockOutUseCase';
import { IUseCase as IDailyRecordFetchRestTimeReasonsUseCase } from '@attendance/domain/useCases/dailyRecord/IFetchRestTimeReasonsUseCase';
import { IUseCase as IDailyRecordSaveFieldsUseCase } from '@attendance/domain/useCases/dailyRecord/ISaveFieldsUseCase';
import { IUseCase as IDailyRecordSaveUseCase } from '@attendance/domain/useCases/dailyRecord/ISaveUseCase';
import { IUseCase as IFixDailyRequestCancelApprovalUseCase } from '@attendance/domain/useCases/fixDailyRequest/ICancelApprovalUseCase';
import { IUseCase as IFixDailyRequestCancelSubmittedUseCase } from '@attendance/domain/useCases/fixDailyRequest/ICancelSubmittedUseCase';
import { IUseCase as IFixDailyRequestSubmitUseCase } from '@attendance/domain/useCases/fixDailyRequest/ISubmitUseCase';
import { IUseCase as ILegalAgreementFetchOvertimeUseCase } from '@attendance/domain/useCases/legalAgreement/IFetchOvertimeUseCase';
import { IUseCase as ILegalAgreementRequestApprovalRequestUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ICancelApprovalUseCase';
import { IUseCase as ILegalAgreementRequestCancelRequestUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ICancelRequestUseCase';
import { IUseCase as ILegalAgreementRequestFetchListUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IFetchListUseCase';
import { IUseCase as ILegalAgreementRequestReapplyUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IReapplyUseCase';
import { IUseCase as ILegalAgreementRequestRemoveUseCase } from '@attendance/domain/useCases/legalAgreementRequest/IRemoveUseCase';
import { IUseCase as ILegalAgreementRequestSubmitUseCase } from '@attendance/domain/useCases/legalAgreementRequest/ISubmitUseCase';
import { IUseCase as IObjectivelyEventLogCreateUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ICreateUseCase';
import { IUseCase as IObjectivelyEventLogFetchDailyDeviationReasonsUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchDailyDeviationReasonsUseCase';
import { IUseCase as IObjectivelyEventLogFetchDailyUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchDailyUseCase';
import { IUseCase as IObjectivelyEventLogFetchUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IFetchUseCase';
import { IUseCase as IObjectivelyEventLogRemoveUseCase } from '@attendance/domain/useCases/objectivelyEventLog/IRemoveUseCase';
import { IUseCase as IObjectivelyEventLogSaveDailyDeviationReasonUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ISaveDailyDeviationReasonUseCase';
import { IUseCase as IObjectivelyEventLogSetToBeAppliedToDailyUseCase } from '@attendance/domain/useCases/objectivelyEventLog/ISetToBeAppliedToDailyUseCase';
import { IUseCase as IStampTimePostUseCase } from '@attendance/domain/useCases/stampTime/IPostUseCase';
import { IUseCase as IDisplayFieldLayoutFetchTableUseCase } from '@attendance/domain/useCases/timesheet/IFetchTableUseCase';
import { IUseCase as ITimesheetFetchUseCase } from '@attendance/domain/useCases/timesheet/IFetchUseCase';

export type UseCases = {
  fetchTimesheet: ITimesheetFetchUseCase;
  reloadTimesheetOnly: ITimesheetFetchUseCase;
  fetchObjectivelyEventLogs: IObjectivelyEventLogFetchUseCase;
  fetchDailyObjectivelyEventLogDeviationReasons: IObjectivelyEventLogFetchDailyDeviationReasonsUseCase;
  setToBeAppliedToDailyObjectivelyEventLog: IObjectivelyEventLogSetToBeAppliedToDailyUseCase;
  reloadDailyObjectivelyEventLogs: IObjectivelyEventLogFetchDailyUseCase;
  reloadOneDailyObjectivelyEventLog: IObjectivelyEventLogFetchDailyUseCase;
  createObjectivelyEventLog: IObjectivelyEventLogCreateUseCase;
  removeObjectivelyEventLog: IObjectivelyEventLogRemoveUseCase;
  submitFixDailyRequest: IFixDailyRequestSubmitUseCase;
  submitFixDailyRequestWithClockOut: ISubmitFixDailyRequestWithClockOutUseCase;
  submitFixDailyRequestAndSaveDailyRecord: ISubmitFixDailyRequestAndSaveDailyRecordUseCase;
  cancelSubmittedFixDailyRequest: IFixDailyRequestCancelSubmittedUseCase;
  cancelApprovalFixDailyRequest: IFixDailyRequestCancelApprovalUseCase;
  saveDailyRecord: IDailyRecordSaveUseCase;
  saveDailyObjectivelyEventLogDeviationReason: IObjectivelyEventLogSaveDailyDeviationReasonUseCase;
  stampTime: IStampTimePostUseCase;
  fetchRestTimeReasons: IDailyRecordFetchRestTimeReasonsUseCase;
  fetchListLegalAgreementRequest: ILegalAgreementRequestFetchListUseCase;
  fetchOvertimeLegalAgreement: ILegalAgreementFetchOvertimeUseCase;
  submitLegalAgreementRequest: ILegalAgreementRequestSubmitUseCase;
  cancelRequestLegalAgreementRequest: ILegalAgreementRequestCancelRequestUseCase;
  cancelApprovalLegalAgreementRequest: ILegalAgreementRequestApprovalRequestUseCase;
  removeLegalAgreementRequest: ILegalAgreementRequestRemoveUseCase;
  reapplyLegalAgreementRequest: ILegalAgreementRequestReapplyUseCase;
  fetchDailyFieldLayoutTable: IDisplayFieldLayoutFetchTableUseCase;
  saveFields: IDailyRecordSaveFieldsUseCase;
};

const service = UseCaseCollection<UseCases>("Timesheet-PC's UseCase Service");

export default service;
