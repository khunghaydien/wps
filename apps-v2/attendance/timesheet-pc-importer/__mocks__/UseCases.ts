import { UseCases } from '../UseCases';

export const service: UseCases = {
  fetchUserSetting: jest.fn(),
  fetchRestTimeReasons: jest.fn(),
  fetchRestTimeReasonsForBulk: jest.fn(),
  fetchLeaves: jest.fn(),
  fetchEarlyLeaveReasons: jest.fn(),
  fetchLateArrivalReasons: jest.fn(),
  fetchContractedWorkTimes: jest.fn(),
  saveTimesheet: jest.fn(),
  checkTimesheet: jest.fn(),
};

export default (): UseCases => service;
