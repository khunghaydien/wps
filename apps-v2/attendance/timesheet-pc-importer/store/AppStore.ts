import { Action, Store } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

import { State } from '@attendance/timesheet-pc-importer/modules';

export type AppState = State;
export type AppStore = Store<State>;
export type AppDispatch = ThunkDispatch<unknown, void, Action<string>>;
