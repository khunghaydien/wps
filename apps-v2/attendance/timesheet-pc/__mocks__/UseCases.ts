import { UseCases } from '../UseCases';

export const methods: UseCases = {
  fetchTimesheet: jest.fn().mockResolvedValue({
    timesheet: 'timesheet',
  }),
  reloadTimesheetOnly: jest.fn().mockResolvedValue({
    timesheet: 'timesheet',
  }),
  fetchObjectivelyEventLogs: jest.fn().mockResolvedValue({
    objectivelyEventLogs: 'objectivelyEventLogs',
  }),
  setToBeAppliedToDailyObjectivelyEventLog: jest
    .fn()
    .mockResolvedValue('result'),
  reloadDailyObjectivelyEventLogs: jest.fn().mockResolvedValue({
    dailyObjectivelyEventLogs: ['dailyObjectivelyEventLog'],
  }),
  reloadOneDailyObjectivelyEventLog: jest.fn().mockResolvedValue({
    dailyObjectivelyEventLogs: ['dailyObjectivelyEventLog'],
  }),
  createObjectivelyEventLog: jest.fn().mockResolvedValue('result'),
  removeObjectivelyEventLog: jest.fn().mockResolvedValue('result'),
  stampTime: jest.fn().mockResolvedValue('result'),
  saveDailyRecord: jest.fn().mockResolvedValue('result'),
  saveDailyObjectivelyEventLogDeviationReason: jest
    .fn()
    .mockResolvedValue('result'),
  submitFixDailyRequest: jest.fn().mockResolvedValue('result'),
  submitFixDailyRequestWithClockOut: jest.fn().mockRejectedValue('result'),
  submitFixDailyRequestAndSaveDailyRecord: jest
    .fn()
    .mockRejectedValue('result'),
  cancelSubmittedFixDailyRequest: jest.fn().mockResolvedValue('result'),
  cancelApprovalFixDailyRequest: jest.fn().mockResolvedValue('result'),
  fetchRestTimeReasons: jest.fn().mockResolvedValue('result'),
  fetchDailyObjectivelyEventLogDeviationReasons: jest
    .fn()
    .mockResolvedValue('result'),
  fetchListLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  fetchOvertimeLegalAgreement: jest.fn().mockResolvedValue('result'),
  submitLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  reapplyLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  cancelRequestLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  cancelApprovalLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  removeLegalAgreementRequest: jest.fn().mockResolvedValue('result'),
  fetchDailyFieldLayoutTable: jest.fn().mockResolvedValue('result'),
  saveFields: jest.fn().mockResolvedValue('result'),
};

export default (): UseCases => methods;
