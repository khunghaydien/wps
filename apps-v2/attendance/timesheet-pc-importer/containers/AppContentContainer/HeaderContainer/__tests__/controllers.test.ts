import configureMockStore from 'redux-mock-store';

import createControllers from '../controllers';
import { AppStore } from '@attendance/timesheet-pc-importer/store/AppStore';
import UseCases from '@attendance/timesheet-pc-importer/UseCases';
import TimesheetFactory from '@attendance/timesheet-pc-importer/viewModels/DailyRecordViewModel/TimesheetFactory';

jest.mock('@attendance/timesheet-pc-importer/modules/selectors', () => ({
  __esModules: true,
  ownerEmployeeId: () => 'ownerEmployeeId',
}));
jest.mock('@attendance/timesheet-pc-importer/UseCases');
jest.mock(
  '@attendance/timesheet-pc-importer/viewModels/DailyRecordViewModel/TimesheetFactory'
);

const mockStore = configureMockStore();

beforeEach(() => {
  jest.clearAllMocks();
});

describe('updateStartDate', () => {
  it('should call', async () => {
    // Arrange
    const store = mockStore({
      timesheet: {
        endDate: '2023-01-31',
      },
    });
    const controllers = createControllers(store as AppStore);

    // Act
    await controllers.updateStartDate('2023-01-01');

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  it.each(['2023-01-01', '2023-03-01', '', null])(
    'should calls with alert',
    async (startDate) => {
      // Arrange
      const store = mockStore({
        timesheet: {
          endDate: '2023-02-28',
        },
      });
      const controllers = createControllers(store as AppStore);

      // Act
      await controllers.updateStartDate(startDate);

      // Assert
      expect(store.getActions()).toMatchSnapshot();
    }
  );
});

describe('updateEndDate', () => {
  it('should call', async () => {
    // Arrange
    const store = mockStore({
      timesheet: {
        startDate: '2023-01-01',
      },
    });
    const controllers = createControllers(store as AppStore);

    // Act
    await controllers.updateEndDate('2023-01-31');

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  it.each(['2023-02-28', '2022-12-31', '', null])(
    'should calls with alert',
    async (endDate) => {
      // Arrange
      const store = mockStore({
        timesheet: {
          startDate: '2023-01-01',
        },
      });
      const controllers = createControllers(store as AppStore);

      // Act
      await controllers.updateEndDate(endDate);

      // Assert
      expect(store.getActions()).toMatchSnapshot();
    }
  );
});

describe('checkTimesheet', () => {
  it('should call', async () => {
    // Arrange
    const store = mockStore({
      timesheet: {
        startDate: 'storeStartDate',
        endDate: 'storeEndDate',
        records: new Map([
          ['recordDate1', { checked: true, recordDate: 'recordDate1' }],
          ['recordDate2', { checked: true, recordDate: 'recordDate2' }],
        ]),
      },
    });
    const controllers = createControllers(store as AppStore);

    // Act
    await controllers.checkTimesheet();

    // Assert
    expect(UseCases().checkTimesheet).toHaveBeenCalledTimes(1);
    expect(UseCases().checkTimesheet).toHaveBeenCalledWith({
      employeeId: 'ownerEmployeeId',
      startDate: 'recordDate1',
      endDate: 'recordDate2',
    });
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('saveTimesheet', () => {
  it('should call', async () => {
    // Arrange
    (TimesheetFactory.create as jest.Mock).mockReturnValue(
      'convertedTimesheet'
    );
    const store = mockStore({
      timesheet: {
        startDate: 'startDate',
        endDate: 'endDate',
        records: new Map([
          ['recordDate1', { recordDate: 'recordDate1' }],
          ['recordDate2', { recordDate: 'recordDate2' }],
        ]),
      },
    });
    const controllers = createControllers(store as AppStore);

    // Act
    await controllers.saveTimesheet();

    // Assert
    expect(TimesheetFactory.create).toHaveBeenCalledTimes(1);
    expect(TimesheetFactory.create).toHaveBeenCalledWith({
      employeeId: 'ownerEmployeeId',
      records: [
        {
          recordDate: 'recordDate1',
        },
        {
          recordDate: 'recordDate2',
        },
      ],
    });
    expect(UseCases().saveTimesheet).toHaveBeenCalledTimes(1);
    expect(UseCases().saveTimesheet).toHaveBeenCalledWith('convertedTimesheet');
    expect(store.getActions()).toMatchSnapshot();
  });
});
