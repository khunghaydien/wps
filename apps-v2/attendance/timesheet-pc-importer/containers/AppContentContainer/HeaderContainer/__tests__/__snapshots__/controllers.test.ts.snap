// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkTimesheet should call 1`] = `Array []`;

exports[`saveTimesheet should call 1`] = `Array []`;

exports[`updateEndDate should call 1`] = `
Array [
  Object {
    "payload": Object {
      "endDate": "2023-01-31",
      "startDate": "2023-01-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateEndDate should calls with alert 1`] = `
Array [
  Object {
    "payload": [Error: Error - The period must not exceed 1 month.
If you want apply after Jan 31, 2023, please change the start date.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "endDate": "2023-02-28",
      "startDate": "2023-01-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateEndDate should calls with alert 2`] = `
Array [
  Object {
    "payload": [Error: Error - The period must not exceed 1 month.
If you want apply after Jan 31, 2023, please change the start date.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "endDate": "2022-12-31",
      "startDate": "2023-01-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateEndDate should calls with alert 3`] = `
Array [
  Object {
    "payload": [Error: Error - The period must not exceed 1 month.
If you want apply after Jan 31, 2023, please change the start date.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "endDate": "",
      "startDate": "2023-01-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateEndDate should calls with alert 4`] = `
Array [
  Object {
    "payload": Object {
      "endDate": null,
      "startDate": "2023-01-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateStartDate should call 1`] = `
Array [
  Object {
    "payload": Object {
      "endDate": "2023-01-31",
      "startDate": "2023-01-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateStartDate should calls with alert 1`] = `
Array [
  Object {
    "payload": [Error: Error - The period must not exceed 1 month.
If you want apply after Jan 31, 2023, please change the start date.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "endDate": "2023-02-28",
      "startDate": "2023-01-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateStartDate should calls with alert 2`] = `
Array [
  Object {
    "payload": [Error: Error - The period must not exceed 1 month.
If you want apply after Mar 31, 2023, please change the start date.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "endDate": "2023-02-28",
      "startDate": "2023-03-01",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateStartDate should calls with alert 3`] = `
Array [
  Object {
    "payload": [Error: Error - The period must not exceed 1 month.
If you want apply after , please change the start date.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "endDate": "2023-02-28",
      "startDate": "",
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;

exports[`updateStartDate should calls with alert 4`] = `
Array [
  Object {
    "payload": [Error: Error - The period must not exceed 1 month.
If you want apply after , please change the start date.],
    "type": "CATCH_BUSINESS_ERROR",
  },
  Object {
    "payload": Object {
      "endDate": "2023-02-28",
      "startDate": null,
    },
    "type": "TIMESHEET-PC-IMPORTER/DAILY_RECORDS/CREATE",
  },
]
`;
